### PR TITLE
fix(react): inline overlays dismiss when parent component unmounts

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

When navigation events occur (push, pop, replace), presented inline overlays are not unmounted from the DOM. Additionally, when a parent component is unmounted, presented overlays are not unmounted. 

Issue URL: #25775, #26185

## What is the new behavior?

- When a parent component is unmounted, any presented overlay within the context of that component will additionally be unmounted.
  - This operation **skips** the dismiss life cycle and does not perform the dismiss transition, emit the `willDismiss` or `didDismiss` events.
- Navigation actions, such as push, pop or replace will unmounted presented overlays.
  - This operation **skips** the dismiss life cycle and does not perform the dismiss transition, emit the `willDismiss` or `didDismiss` events.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Dev-build: `6.5.4-dev.11676062329.117ad0d6` (updated 02/10/2023)

### CODE REVIEW

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 1]
```diff
@@ -105,6 +105,12 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
       attachProps(node, this.props, prevProps);
     }
 
+    componentWillUnmount() {
+      if (this.ref.current) {
+        this.ref.current.dismiss();
```
Comment (amandaejohnston) [inline packages/react/src/components/createInlineOverlayComponent.tsx:110]: I'm getting a console error about changing state on an unmounted component when using the Replace or Push operations. I'm guessing `dismiss` is trying to update things after the modal has already been unmounted; maybe it needs to be `await`ed? 🤔 

Test repo/branch I'm using (need to install the dev build): https://github.com/amandaejohnston/sandbox-ionic-react/tree/FW-1889
Screencast: https://user-images.githubusercontent.com/90629384/200869178-3e9a8130-c1b9-42be-84b7-d870bd518887.mp4

**Review by amandaejohnston — CHANGES_REQUESTED** [review]

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 1 above)
Comment (liamdebeasi) [reply packages/react/src/components/createInlineOverlayComponent.tsx:110]: `componentWillUnmount` is synchronous, so `await` won't do anything here. We might need to add logic here to only update the state if the component is not actively unmounting: https://github.com/ionic-team/ionic-framework/blob/1ea0cd29d93eb29561c016c215e79f182e07ce8b/packages/react/src/components/createInlineOverlayComponent.tsx#L66-L99 (I'm guessing this is where the state update is coming from, but it could be somewhere else too)

Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:110]: This issue occurs because the `onDidDismiss` fires and the `setState` in the app implementation updates after the component unmounts.

This resolves it in implementations: https://github.com/ionic-team/ionic-framework/pull/26245/files#diff-6c9571ce83a153612a7dcb0a9ced4fb6eaf6f3a613c9e7d4c614136311be6173R14

For a Framework fix, we could look at not handling `onDidDismiss` if the component has already unmounted.

Comment (amandaejohnston) [reply packages/react/src/components/createInlineOverlayComponent.tsx:110]: If it's possible to handle this in the framework, I would rather go that route, since otherwise this could be considered a breaking change 🤔

**Comment by sean-perkins** [comment]
Updated dev-build: `6.3.6-dev.11668106431.125dbfa9`

**Review by amandaejohnston — APPROVED** [review]

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 2]
```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
           this.setState({ isOpen: false });
         }
 
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
+        const didComponentUnmount = this.ref.current === null;
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:117]: Optional: Would it be better to remove the `didDismiss` and `willPresent` listeners in `componentWillUnmount` instead?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 3]
```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
+      const propsToPass = Object.keys(this.props).reduce((acc, name) => {
+        if (!excludedOverlayEvents.includes(name)) {
+          /**
+           * Maintains all props except for the event handlers
+           * that are manually handled in the componentDidMount.
+           */
+          (acc as any)[name] = (this.props as any)[name];
+        }
+        return acc;
+      }, {});
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:141]: Why is this needed?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 4]
```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
+      attachProps(node, propsToPass, prevProps);
+    }
+
+    componentWillUnmount() {
+      if (this.ref.current) {
+        this.ref.current.dismiss();
+      }
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:148]: If React handles the rendering of the Web Component, won't the `ion-modal` element be removed automatically?

Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:148]: React will only remove the DOM node if the overlay is dismissed. When the modal is presented, we are manually moving the node to the root and React is unable to know to remove it when the React component unmounts. This check/code moves the node back in time for React to remove it.

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 2 above)
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:117]: Investigated this path, results in the previous error:
> react-dom.development.js:67 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

If I remove the `didComponentUnmount` check even when removing the event listeners in `componentWillUnmount`.

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 3 above)
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:141]: This change prevents `onDidDismiss` being invoked after the component is unmounted, resulting in the React error:
> Can't perform a React state update on an unmounted component.

There are multiple conditions that cause this, from either us trying to handle the method in the event listener after the component has unmounted (`didComponentUnmount` prevents this) or from the `attachProps` util (which also adds an event listener and can execute after the component has unmounted).

**Comment by sean-perkins** [comment]
Thinking about this approach a little more, if we do not like the filtering out of events to pass to `attachProps`. We could probably come up with a `detachProps` utility that pulls the cached events from the `node.__events` event store and remove the event listener when the component is unmounted.

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 5]
```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
+        const didComponentUnmount = this.ref.current === null;
+
+        if (didComponentUnmount === false) {
+          /**
+           * If the component is unmounted, we do not
+           * want to call the onDidDismiss handler to
+           * prevent a React state update on an unmounted component.
+           */
+          this.props.onDidDismiss && this.props.onDidDismiss(evt);
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:125]: Can we add a comment that explains why we don't remove the event listener in the on unmount method?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 6]
```diff
@@ -107,13 +114,38 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
+        }
       });
     }
 
     componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
       const node = this.ref.current! as HTMLElement;
-      attachProps(node, this.props, prevProps);
+      const propsToPass = Object.keys(this.props).reduce((acc, name) => {
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:132]: Optional: It might be better to iterate through the object keys and build a new object as that will only interate through the props once. The proposed approach will iterate through the props twice: One iteration when calling `Object.keys` and then another iteration when calling `reduce`.

**Comment by puopg** [comment]
Just checking in on this fix, could the team provide an update what is its status?

**Comment by patdx** [comment]
I just tried out the dev build `6.3.6-dev.11668106431.125dbfa9` above, it also gives better behavior for React Error Boundary:

```
const BuggyComponent = () => {
  throw new Error("Failed to render!");
};

const App = () => {
  return (
    <ErrorBoundary>
      <IonModal>
        <BuggyComponent />
      </IonModal>
    </ErrorBoundary>
  );
};
```

In the current prod build I just discovered that code like above freezes the (inactive) modal over the whole screen.

**Comment by puopg** [comment]
@sean-perkins Bump?

**Comment by sean-perkins** [comment]
Open PRs are considered active work. There are a number of factors that effects our prioritization of different work items. 

This specific PR addresses behavior that has not been designed or documented by Ionic and is important that we spent time to thoroughly designing the expected behaviors internally, so we do not need to perform a breaking change to correct that behavior in the future. 

This PR has been updated to remove incorrect assumptions I had made in the first implementation pass, such as trying to perform the dismiss transition for the overlay when the parent component unmounts or a router action is performed. In these scenarios, we should be treating the overlay lifecycle as:

```mermaid
graph TD
A[Overlay]--> Mounted
Mounted --> Presented
Presented --> Unmounted
```

In this diagram we show that the "Dismiss" lifecycle does not occur. This means that a dismiss transition is not performed and the respective events `willDismiss` and `didDismiss` are not emitted. This is intentional to avoid memory leaks and unexpected scenarios where callback events are firing as a result of side-effects. 

Here is an updated dev-build to test with: `6.5.4-dev.11676056919.1a0af097`.

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 7]
```diff
@@ -172,6 +151,46 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
     static get displayName() {
       return displayName;
     }
+
+    private handleIonMount = () => {
+      /**
+       * Mount the inner component when the
+       * overlay is about to open.
+       *
+       * For ion-popover, this is when `ionMount` is emitted.
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:160]: What does only the popover emit this? Related to popover positioning?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 8]
```diff
@@ -54,65 +55,43 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
-        }
-
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
-      });
-    }
-
-    componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
-      const node = this.ref.current! as HTMLElement;
-      attachProps(node, this.props, prevProps);
+        node.removeEventListener('didDismiss', this.handleDidDismiss);
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:91]: Why do we only remove the `didDismiss` event?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 9]
```diff
@@ -54,65 +55,43 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
+        node.remove();
```
Comment (liamdebeasi) [inline packages/react/src/components/createInlineOverlayComponent.tsx:92]: Why do we need to manually remove the DOM node? Shouldn't React handle this for us?

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 7 above)
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:160]: Correct, we ran into an issue with popover implementations with auto widths: https://github.com/ionic-team/ionic-framework/pull/26274.

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 8 above)
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:91]: The `createInlineOverlayComponent` binds two internal event listeners for `didDismiss` and `willPresent`. Both implementations have an internal state update for:

```ts
this.setState({ isOpen: false }); // `true` for willPresent
```

At this point in the lifecycle of the component the component is unmounting. `willPresent` cannot fire. However, `didDismiss` can fire and cause any user-implemented callback handlers for `onDidDismiss` to run after the React component has unmounted. If this occurs, React will throw an exception in the console about a memory leak. 

This code disconnects the internal event listener manually, before the element can dispatch `didDismiss`.

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 9 above)
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:92]: At this point in the execution/logic, the overlay is presented and relocated to the root of the DOM. It is outside of the context of React's VDOM to unmount. 

We could try teleporting the node back to have React also remove it, but we would need to also repaint the overlay as dismissed, otherwise it would cause visual overlapping issues with moving a presented overlay back into the nested DOM tree. 

I opted for `node.remove()` since it cleans up the presented overlay and doesn't require any style recalculations or visual changes. We unfortunately are dealing with a limited execution window with `componentWillUnmount`, so we cannot do a blocking task like `await node.dismiss()`.

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 10]
```diff
@@ -172,6 +151,46 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
+      this.props.onWillPresent && this.props.onWillPresent(evt);
+    };
+
+    private handleDidDismiss = (evt: any) => {
+      const wrapper = this.wrapperRef.current;
+      const el = this.ref.current;
+
+      /**
+       * This component might be unmounted already, if the containing
+       * element was removed while the popover was still open. (For
```
[the remark this answers is not in the document]
Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:183]: This comment was just relocated in this commit, prior it came from: https://github.com/ionic-team/ionic-framework/pull/24569

@amandaejohnston do you remember if this is specific to popover or applies to all overlays?

File: packages/react/src/components/createInlineOverlayComponent.tsx [hunk 11]
```diff
@@ -54,65 +55,43 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
...
-        const wrapper = this.wrapperRef.current;
-        const el = this.ref.current;
-
+      if (node && this.state.isOpen) {
         /**
-         * This component might be unmounted already, if the containing
-         * element was removed while the popover was still open. (For
-         * example, if an item contains an inline popover with a button
-         * that removes the item.)
+         * Detach the local event listener than performs the state updates,
```
Comment (amandaejohnston) [inline packages/react/src/components/createInlineOverlayComponent.tsx:86]:
```suggestion
         * Detach the local event listener that performs the state updates,
```

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (amandaejohnston) [reply packages/react/src/components/createInlineOverlayComponent.tsx:183]: In theory this should apply to all overlays; my comment was overly specific 😆

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 8 above)
Comment (amandaejohnston) [reply packages/react/src/components/createInlineOverlayComponent.tsx:91]: Would `detachProps` handle removing that listener? If so, would it make sense to instead call `detachProps` before `node.remove` and then remove this line?

Comment (sean-perkins) [reply packages/react/src/components/createInlineOverlayComponent.tsx:91]: `detachProps` only removes event listeners added by `attachProps` (because we track the events on the element node on a key called `__events`). This specific event listener is manually added within the React class and isn't removed as a result of `detachProps`.

File: packages/react/src/components/createInlineOverlayComponent.tsx (hunk 9 above)
Comment (liamdebeasi) [reply packages/react/src/components/createInlineOverlayComponent.tsx:92]: oh that's right. thanks!
